docs/en/writing-running-appium/ios/ios-xctest-file-movement.md

```diff
@@ -0,0 +1,130 @@
+3. `<path_to_the_file_or_folder_inside_container>`
+
+### Real device
+
+For real devices, Appium relies on [ifuse](https://github.com/libimobiledevice/ifuse) to achieve the feature.
+You may compile ifuse from the source code or via brew.
+
+```
+$ brew cask install osxfuse
+$ brew install ifuse --HEAD # newer iOS versions need the latest codebase
+```
+
+#### Format
+
+The format of method argument should be the following:
+
+- `@<app_bundle_id>` is a bundle id
+- `optional_container_type` is container type
+    - `documents` is only available
+        - BundleIds which can get by `ifuse -u <udid> --list-apps` can specify `documents` type
+        - e.g. Below _On My iPhone_ image has _Slack_ folder, but `com.tinyspeck.chatlyio` does not exist in the output of `--list-apps`. Thus, we cannot mount it as `com.tinyspeck.chatlyio@documents/appium.png`
+    - The others works as _format 2_
+        - Can be mounted apps which have `UIFileSharingEnabled` flag in its `info.plist`
+- `path_to_the_file_or_folder_inside_container` is the target to push/pull to/from them.
+    - If the `optional_container_type` is `documents`, this path will be mapped to
+      `On My iPhone/<app name>` in Files app
+
+_format 3_ is not allowed for real devices.
+
+#### Example
+
+If you would like to pull _Presentation.key_ form Keynote app, you can get it as below.
+
+- Pull file
+
+```javascript
+// webdriver.io
+let data = driver.pullFile('@io.appium.example:documents/Presentation.key');
+```
+
+```ruby
+# ruby_lib_core
+file = @driver.pull_file '@com.apple.Keynote:documents/Presentation.key'
+File.open('presentation.key', 'wb') { |f| f<< file }
+```
+
+The file is in _On My iPhone/Keynote_ on Files app.
+
+|Top | On  My iPhone | Keynote |
+|:----:|:----:|:----:|
+|![](./ios-xctest-file-movement/top_files.png)|![](./ios-xctest-file-movement/on_my_iphone.png)|![](./ios-xctest-file-movement/keynote.png)|
+
+If the file is in deeper place like _On My iPhone/Keynote/Dir1/Dir2_, then the Ruby command should be:
+
+```javascript
+// webdriver.io
+let data = driver.pullFile('@io.appium.example:documents/Dir1/Dir2/Presentation.key');
+```
+
+```ruby
+# ruby_lib_core
+file = @driver.pull_file '@com.apple.Keynote:documents/Dir1/Dir2/Presentation.key'
+File.open('presentation.key', 'wb') { |f| f<< file }
+```
+
+- Pull folder
+
+You can pull root of _On My iPhone/Keynote_ as `@driver.pull_folder '@com.apple.Keynote:documents/'`.
+
+```javascript
+// webdriver.io
+let data = driver.pullFolder('@io.appium.example:documents/');
+```
+
+```ruby
+# ruby_lib_core
+file = @driver.pull_folder '@com.apple.Keynote:documents/'
+```
+
+- Push file
+
+As same as pull files
+
+```javascript
+// webdriver.io
+driver.pushFile('@com.apple.Keynote:documents/text.txt', new Buffer("Hello World").toString('base64'));
+```
+
+```ruby
+# ruby_lib_core
+@driver.push_file '@com.apple.Keynote:documents/text.txt', (File.read 'path/to/file')
```

Comment:
is base64 encoding done inside push_file in ruby lib?

Reply:
Yes.
Ruby has two encodings, `Base64.strict_encode64` and `Base64.encode64`. I found in some case encoded by `encode64` failed to decode correctly by other languages since it has `\n` in some places. To reduce such error cases, Ruby client calls strict one inside the method